### ISSUE TITLE
[Ruby] Segmentation Fault

### ISSUE THREAD

**Issue by tonobo** [google/protobuf#2952]
I've catched a ruby segmentation fault during rspec load.

```
-- C level backtrace information -------------------------------------------
/usr/local/lib/libruby.so.2.4(rb_vm_bugreport+0x4f3) [0x7f3517b1d683] vm_dump.c:684
/usr/local/lib/libruby.so.2.4(rb_bug_context+0xd4) [0x7f351799c3a4] error.c:506
/usr/local/lib/libruby.so.2.4(sigsegv+0x3e) [0x7f3517a9105e] signal.c:907
/lib/x86_64-linux-gnu/libpthread.so.0 [0x7f35176e8890]
/usr/local/lib/libruby.so.2.4(gc_mark_ptr+0x69) [0x7f35179b8609] gc.c:1111
/usr/local/lib/libruby.so.2.4(gc_marks_rest+0x10d) [0x7f35179bb7ad] gc.c:4629
/usr/local/lib/libruby.so.2.4(gc_start.part.65+0x48d) [0x7f35179bbdbd] gc.c:5606
/usr/local/lib/libruby.so.2.4(heap_get_freeobj_from_next_freepage+0xdf) [0x7f35179be4af] gc.c:1696
/usr/local/lib/libruby.so.2.4(newobj_slowpath_wb_protected+0x88) [0x7f35179be938] gc.c:1742
/usr/local/lib/libruby.so.2.4(rb_ary_new_capa+0x35) [0x7f351792b055] array.c:445
/usr/local/bundle/gems/google-protobuf-3.2.0.2-x86_64-linux/lib/google/2.4/protobuf_c.so(Builder_alloc+0x31) [0x7f3512a845e1]
/usr/local/lib/libruby.so.2.4(rb_obj_alloc+0x3a) [0x7f3517a0e3ca] object.c:1863
/usr/local/lib/libruby.so.2.4(rb_class_new_instance+0x11) [0x7f3517a0e7e1] object.c:1895
/usr/local/bundle/gems/google-protobuf-3.2.0.2-x86_64-linux/lib/google/2.4/protobuf_c.so(DescriptorPool_build+0x21) [0x7f3512a84291]
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x113f) [0x7f3517b08eff] insns.def:967
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(rb_load_internal0+0xaf) [0x7f35179e7dbf] load.c:617
/usr/local/lib/libruby.so.2.4(rb_require_internal+0x630) [0x7f35179e98d0] load.c:997
/usr/local/lib/libruby.so.2.4(rb_require_safe+0x9) [0x7f35179e9c49] load.c:1042
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x1610) [0x7f3517b093d0] insns.def:1066
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(rb_load_internal0+0xaf) [0x7f35179e7dbf] load.c:617
/usr/local/lib/libruby.so.2.4(rb_require_internal+0x630) [0x7f35179e98d0] load.c:997
/usr/local/lib/libruby.so.2.4(rb_require_safe+0x9) [0x7f35179e9c49] load.c:1042
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x1610) [0x7f3517b093d0] insns.def:1066
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(rb_load_internal0+0xaf) [0x7f35179e7dbf] load.c:617
/usr/local/lib/libruby.so.2.4(rb_require_internal+0x630) [0x7f35179e98d0] load.c:997
/usr/local/lib/libruby.so.2.4(rb_require_safe+0x9) [0x7f35179e9c49] load.c:1042
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x1610) [0x7f3517b093d0] insns.def:1066
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(rb_load_internal0+0xaf) [0x7f35179e7dbf] load.c:617
/usr/local/lib/libruby.so.2.4(rb_require_internal+0x630) [0x7f35179e98d0] load.c:997
/usr/local/lib/libruby.so.2.4(rb_require_safe+0x9) [0x7f35179e9c49] load.c:1042
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x1610) [0x7f3517b093d0] insns.def:1066
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(rb_load_internal0+0xaf) [0x7f35179e7dbf] load.c:617
/usr/local/lib/libruby.so.2.4(rb_require_internal+0x630) [0x7f35179e98d0] load.c:997
/usr/local/lib/libruby.so.2.4(rb_require_safe+0x9) [0x7f35179e9c49] load.c:1042
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x1610) [0x7f3517b093d0] insns.def:1066
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(rb_load_internal0+0xaf) [0x7f35179e7dbf] load.c:617
/usr/local/lib/libruby.so.2.4(rb_require_internal+0x630) [0x7f35179e98d0] load.c:997
/usr/local/lib/libruby.so.2.4(rb_require_safe+0x9) [0x7f35179e9c49] load.c:1042
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x1610) [0x7f3517b093d0] insns.def:1066
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(rb_load_internal0+0xaf) [0x7f35179e7dbf] load.c:617
/usr/local/lib/libruby.so.2.4(rb_require_internal+0x630) [0x7f35179e98d0] load.c:997
/usr/local/lib/libruby.so.2.4(rb_require_safe+0x9) [0x7f35179e9c49] load.c:1042
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x1610) [0x7f3517b093d0] insns.def:1066
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(rb_load_internal0+0xaf) [0x7f35179e7dbf] load.c:617
/usr/local/lib/libruby.so.2.4(rb_f_load+0x8b) [0x7f35179e856b] load.c:646
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x1610) [0x7f3517b093d0] insns.def:1066
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(invoke_iseq_block_from_c+0x516) [0x7f3517b0f206] vm.c:969
/usr/local/lib/libruby.so.2.4(rb_yield+0xbd) [0x7f3517b0f8ad] vm.c:1032
/usr/local/lib/libruby.so.2.4(rb_ary_each+0x3d) [0x7f351792723d] array.c:1824
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x113f) [0x7f3517b08eff] insns.def:967
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(rb_load_internal0+0xaf) [0x7f35179e7dbf] load.c:617
/usr/local/lib/libruby.so.2.4(rb_f_load+0x8b) [0x7f35179e856b] load.c:646
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x1610) [0x7f3517b093d0] insns.def:1066
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(rb_load_internal0+0xaf) [0x7f35179e7dbf] load.c:617
/usr/local/lib/libruby.so.2.4(rb_f_load+0x8b) [0x7f35179e856b] load.c:646
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x1610) [0x7f3517b093d0] insns.def:1066
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(rb_load_internal0+0xaf) [0x7f35179e7dbf] load.c:617
/usr/local/lib/libruby.so.2.4(rb_f_load+0x8b) [0x7f35179e856b] load.c:646
/usr/local/lib/libruby.so.2.4(vm_call_cfunc+0xef) [0x7f3517afff3f] vm_insnhelper.c:1752
/usr/local/lib/libruby.so.2.4(vm_call_method+0xe3) [0x7f3517b11b13] vm_insnhelper.c:2292
/usr/local/lib/libruby.so.2.4(vm_exec_core+0x1610) [0x7f3517b093d0] insns.def:1066
/usr/local/lib/libruby.so.2.4(vm_exec+0x8b) [0x7f3517b0e42b] vm.c:1727
/usr/local/lib/libruby.so.2.4(ruby_exec_internal+0xb2) [0x7f35179a2132] eval.c:244
/usr/local/lib/libruby.so.2.4(ruby_exec_node+0x1d) [0x7f35179a3ead] eval.c:308
/usr/local/lib/libruby.so.2.4(ruby_run_node+0x1e) [0x7f35179a6a6e] eval.c:300
/usr/local/bin/ruby(main+0x4b) [0x40087b] main.c:36
```

**Comment by acozzette** [comment]
@dopykuh Could you provide some more details on how to reproduce this? Does the segmentation fault occur reliably or does it just happen randomly at times?

**Comment by dopykuh** [comment]
I am unable to reproduce it, it just happend randomly

**Comment by TeBoring** [comment]
Fixed in #3560.
